README.md
# http-server-pwa aka pwa-server

[![npm](https://img.shields.io/npm/v/http-server-pwa.svg?style=flat)](https://github.com/LasaleFamine/http-server-pwa) [![Build Status](https://travis-ci.com/LasaleFamine/http-server-pwa.svg?branch=master&style=flat)](https://travis-ci.com/LasaleFamine/http-server-pwa) [![Build status](https://ci.appveyor.com/api/projects/status/k5ssy06tt4ru1269?svg=true&style=flat)](https://ci.appveyor.com/project/LasaleFamine/http-server-pwa) [![codecov](https://codecov.io/gh/LasaleFamine/http-server-pwa/badge.svg?branch=master&style=flat)](https://codecov.io/gh/LasaleFamine/http-server-pwa?branch=master)

> http-server alike but for serving and rendering PWA: pwa-server

## What

This is a different version (btw inspired by) of the insta-ready [http-server](https://github.com/indexzero/http-server), but with some key features to render Progressive Web Apps (or also SPA) correctly for bots like `googlebot` or Facebook crawlers, so your application can be crawled with its content already loaded. This should avoid the killing problem of the SEO for PWAs.

**The server will render as normal, static ExpressJS server for your normal users.**

Worth noting is that the server supports the `history-fallback` behavior. You maybe can avoid that ***#*** (hash) for your PWA navigation ✌️.

## Features

The server is not "dependecies-free" like the original http-server and instead is powered by [ExpressJS](https://github.com/expressjs/express) under the hood and [Puppeter](https://github.com/GoogleChrome/puppeteer/) for render the pages to bots.

* [pupperender](https://github.com/LasaleFamine/pupperender) - Puppeter middleware to render correctly the PWA content for the crawlers.
* [express-history-api-fallback](https://www.npmjs.com/package/express-history-api-fallback) - to fallback correctly to the `index.html`.
* [express-http-to-https](https://www.npmjs.com/package/express-http-to-https) - automatically redirect `http` requests to `https`.
* [devcert](https://github.com/davewasmer/devcert) - Create development certificate on the fly for local `https`. **NOTE: you could be prompted to insert your password before starting the server. This is necessary of using OpenSSL. More info on devcert repository.**

## Production usage

The server will not create any SSL certificate on the fly when you set `process.env.NODE_ENV = production` even if `--ssl` flag is passed down.
You have two alternatives here:
- use a reverse proxy like [CloudFlare](https://cloudflare.com) and get free certification for your domain - easy peasy
- use a custom reverse proxy like Nginx and load your certification from there

I currently don't want to support a custom certificate load. PR are always welcome.

## Deploy on Heroku

In order to make the server work as expected on Heroku services (even on free tier) you must add [the puppeteer-heroku-buildpack](https://github.com/jontewks/puppeteer-heroku-buildpack).

Since recent changes on the platform it is **strongly** suggested to add it from source rather than the buildpack's catalogs.

```bash
$ heroku buildpacks:set https://github.com/jontewks/puppeteer-heroku-buildpack.git
```

## Usage

You can use it programmatically or as a [CLI](#CLI) tool (global or local).

## Install

```
$ yarn add http-server-pwa
```

## Usage

```js
const httpServerPwa = require('http-server-pwa');

const server = await httpServerPwa('./dist', {p: 3000});
//=> Server started -> ./dist localhost:3000
```

## API

### httpServerPwa(path, [options])

#### path

Type: `string`

Path to serve.

#### options

| Name | CLI flag | Type | Default | Description |
|------|--------------|------|---------|-------------|
| port | p | `number` | `8080` | Port to use for running the server. |
| host | h | `string` | `localhost` | Host to use for running the server. |
| fallback | f | `string` | `index.html` | Fallback HTML file name. |
| https | s | `boolean` | `false` | Enable HTTPS redirect on `localhost`.|
| ssl |  | `boolean` | `false` | Enable `devcert` auto-generation of SSL certification for development.|
| cache | c | `boolean` | `false` | Enable cache for Puppeteer rendering|
| cacheTTL |  | `number` | `3600` (s) | Seconds until cached content is disregarded and puppeterized again. Only considered when `cache` is true.|
| debug | d | `boolean` | `false` | If or not show some logs.|
| gzip | g | `boolean` | `false` | Enable serving of gzipped files if available|
| brotli | b | `boolean` | `false` | Enable serving of brotli compressed files if available|

> Note: If the gzip and brotli option are both toggled, brotli compressed files will get precedence over gzipped files.

## CLI

```
$ yarn add --global http-server-pwa
```

```
$ http-server-pwa --help # or pwa-server --help

  Usage
	  $ http-server-pwa [path] [options]

	Options
		-p --port       Port to use [Default: 8080]
		-h --host       Host to use [Default: localhost | Windows: 127.0.0.1]
		-f --fallback   Fallback HTML file name [Default: index.html]
		-s --https      Enable HTTPS redirect on localhost [Default: false]
		--ssl           Auto-generation SSL certificate during development [Default: false]
		-c --cache      Enable cache for Puppeteer rendering [Default: false]
		--cacheTTL      Seconds until cached content is disregarded and puppeterized again [Default: 3600 (s)].
		-d --debug      Be more verbose [Default: false]
		-g --gzip	Enable serving of gzipped files if available [Default: false]
		-b --brotli	Enable serving of brotli compressed files if available [Default: false]
		-h --help       Show this message

	Examples
	  $ http-server-pwa
	  Server started -> ./ localhost:8080
	  $ http-server-pwa dist -p 3000
	  Server started -> ./dist localhost:3000
```


## License

MIT © [LasaleFamine](https://godev.space)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.1.0](https://github.com/LasaleFamine/http-server-pwa/compare/v2.0.1...v2.1.0) (2020-10-15)


### Bug Fixes

* indentation ([eb540d2](https://github.com/LasaleFamine/http-server-pwa/commit/eb540d298340969fe44eb65defddda2370e96048))

### [2.0.1](https://github.com/LasaleFamine/http-server-pwa/compare/v2.0.0...v2.0.1) (2020-10-01)


### Bug Fixes

* upgrade chalk from 4.0.0 to 4.1.0 ([e33db50](https://github.com/LasaleFamine/http-server-pwa/commit/e33db5055fdb362cf1b7cffe1629bed1d3070322))
* upgrade meow from 6.1.0 to 6.1.1 ([797e946](https://github.com/LasaleFamine/http-server-pwa/commit/797e94688ce18a38d3393faf51dadcd48c2b7b69))
* upgrade meow from 6.1.0 to 6.1.1 ([de5b58f](https://github.com/LasaleFamine/http-server-pwa/commit/de5b58f778c3bdc226f21d63c94abaa15826449b))

## [2.0.0](https://github.com/LasaleFamine/http-server-pwa/compare/v1.0.1...v2.0.0) (2020-04-06)


### Bug Fixes

* **package:** update chalk to version 2.4.2 ([813c663](https://github.com/LasaleFamine/http-server-pwa/commit/813c663ce72fb9e04debafdcd3a0a11f5ce0abe1))

<a name="1.0.1"></a>
## [1.0.1](https://github.com/LasaleFamine/http-server-pwa/compare/v1.0.0...v1.0.1) (2019-01-04)


### Bug Fixes

* **package:** update pupperender to version 1.2.1 ([c96094d](https://github.com/LasaleFamine/http-server-pwa/commit/c96094d))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/LasaleFamine/http-server-pwa/compare/v0.3.0...v1.0.0) (2018-10-11)


### Features

* exposing cache and cacheTTL from pupperender ([731ab08](https://github.com/LasaleFamine/http-server-pwa/commit/731ab08))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/LasaleFamine/http-server-pwa/compare/v0.2.1...v0.3.0) (2018-09-26)


### Features

* support for devcert enabling SSL auto-generation on dev ([700c44a](https://github.com/LasaleFamine/http-server-pwa/commit/700c44a))
* support for SSL cert on the fly on development and pem for prod ([703b978](https://github.com/LasaleFamine/http-server-pwa/commit/703b978))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/LasaleFamine/http-server-pwa/compare/v0.2.0...v0.2.1) (2018-09-26)


### Bug Fixes

* switch to express-http-to-https for preserving the port ([f031358](https://github.com/LasaleFamine/http-server-pwa/commit/f031358))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/LasaleFamine/http-server-pwa/compare/v0.1.4...v0.2.0) (2018-09-25)


### Bug Fixes

* **package:** update pupperender to version 1.2.0 ([1a2d5bf](https://github.com/LasaleFamine/http-server-pwa/commit/1a2d5bf))


### Features

* you can call it "pwa-server" (thanks [@moebiusmania](https://github.com/moebiusmania) for the idea) ([970bce2](https://github.com/LasaleFamine/http-server-pwa/commit/970bce2))



<a name="0.1.4"></a>
## [0.1.4](https://github.com/LasaleFamine/http-server-pwa/compare/v0.1.3...v0.1.4) (2018-05-09)


### Bug Fixes

* **package:** update chalk to version 2.3.1 ([ed350e3](https://github.com/LasaleFamine/http-server-pwa/commit/ed350e3))
* **package:** update chalk to version 2.3.2 ([905e609](https://github.com/LasaleFamine/http-server-pwa/commit/905e609))
* **package:** update chalk to version 2.4.0 ([5f398a2](https://github.com/LasaleFamine/http-server-pwa/commit/5f398a2))
* **package:** update meow to version 4.0.1 ([b4a08c6](https://github.com/LasaleFamine/http-server-pwa/commit/b4a08c6))
* **package:** update pupperender to version 1.1.1 ([9c3da01](https://github.com/LasaleFamine/http-server-pwa/commit/9c3da01)), closes [#16](https://github.com/LasaleFamine/http-server-pwa/issues/16)



<a name="0.1.3"></a>
## [0.1.3](https://github.com/LasaleFamine/http-server-pwa/compare/v0.1.2...v0.1.3) (2018-01-17)


### Bug Fixes

* **package:** update pupperender to version 1.0.0 ([aa9aa4c](https://github.com/LasaleFamine/http-server-pwa/commit/aa9aa4c))



<a name="0.1.2"></a>
## [0.1.2](https://github.com/LasaleFamine/http-server-pwa/compare/v0.1.1...v0.1.2) (2017-12-30)


### Bug Fixes

* **package:** update meow to version 4.0.0 ([a9a5169](https://github.com/LasaleFamine/http-server-pwa/commit/a9a5169))
* **package:** update pupperender to version 0.1.4 ([199bbde](https://github.com/LasaleFamine/http-server-pwa/commit/199bbde))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/LasaleFamine/http-server-pwa/compare/v0.1.0...v0.1.1) (2017-11-04)


### Bug Fixes

* **pupperender:** update pupperender after fix for Chrome headless ([b75fb27](https://github.com/LasaleFamine/http-server-pwa/commit/b75fb27))



<a name="0.1.0"></a>
# 0.1.0 (2017-11-03)


test/helpers.js
const {resolve} = require('path');
const supertest = require('supertest');
const getPort = require('get-port');

const httpServer = require('./../src');

/**
 * Create the server.
 * @param {!Object} app The app.
 * @return {Promise<object>} Promise of the Server instance and the URL.
 */
exports.listen = async (folder, options) => {
	const correctFolder = folder ? resolve(__dirname, folder) : null;
	const server = await httpServer(correctFolder, {...options});
	return {
		server,
		url: `${options.ssl ? 'https://' : ''}${options.h}:${server.address().port}`
	};
};

/**
 * GET a URL with the given user agent.
 * @param {string} userAgent The user agent string.
 * @param {string} host The host part of the URL.
 * @param {string} path The path part of the URL.
 * @return {Promise<!Object>} Promise of the GET response.
 */
exports.get = (userAgent, host, path) =>
	supertest(host).get(path).set('User-Agent', userAgent);

exports.getWithEncoding = (userAgent, host, path) =>
	supertest(host).get(path).set('User-Agent', userAgent).set('Accept-Encoding', 'br, gzip, deflate, identity');

exports.port = () => getPort();


test/index.test.js
'use strict';

const test = require('ava');
// eslint-disable-next-line ava/no-import-test-files
const {listen, get, getWithEncoding, port} = require('./helpers');

const human = 'Chrome';
const bot = 'slackbot';

test('should work with default params (./ 0.0.0.0:8080)', async t => {
	const {server, url} = await listen(null, {h: 'localhost'});
	const response = await get(human, url, '/');
	// t.is(response.status, 404, 'Not found because there is no index.html');
	t.true(response.text.includes('Cannot GET'), 'Correct Cannot GET on no index.html');
	server.close();
});

test('should render correctly from folder', async t => {
	const {server, url} = await listen('./fixture', {h: 'localhost', p: await port()});
	const response = await get(human, url, '/');
	t.is(response.text, 'some\n');
	server.close();
});

// test('should fallback correctly to index.html', async t => {
// 	const {server, url} = await listen('./fixture', {h: 'localhost', p: await port()});
// 	const response = await get(human, url, '/not-found');
// 	t.is(response.text, 'some\n');
// 	server.close();
// });

test('works with specified host', async t => {
	const customPort = await port();
	const {server, url} = await listen('./fixture', {h: '127.0.0.1', p: customPort});
	const response = await get(human, url, '/');
	t.is(url, `127.0.0.1:${customPort}`);
	t.is(response.text, 'some\n');
	server.close();
});

test('works with different fallback specified', async t => {
	const {server, url} = await listen(
		'./fixture',
		{h: 'localhost', p: await port(), f: 'different.html'}
	);
	const response = await get(human, url, '/something');
	t.is(response.text, 'different fallback\n');
	server.close();
});

test('https redirect by default on not-localhost (with port and path specified)', async t => {
	const customPort = await port();
	const {server, url} = await listen('./fixture',	{h: '0.0.0.0', p: customPort});
	const response = await get(human, url, '/something');
	t.is(response.text, `Found. Redirecting to https://0.0.0.0:${customPort}/something`);
	t.is(response.status, 302);
	server.close();
});

test('https redirect also on localhost', async t => {
	const customPort = await port();
	const {server, url} = await listen('./fixture',	{h: 'localhost', p: customPort, s: true});
	const response = await get(human, url, '/something');
	t.is(response.text, `Found. Redirecting to https://localhost:${customPort}/something`);
	t.is(response.status, 302);
	server.close();
});

test('cache results works correctly', async t => {
	const {server, url} = await listen('./fixture',	{h: 'localhost', p: await port(), cache: true});
	const response = await get(bot, url, '/something');
	t.is(response.status, 200);
	t.false(Boolean(response.get('Expires')));
	const cachedResponse = await get(bot, url, '/something');
	t.is(cachedResponse.status, 200);
	t.true(Boolean(cachedResponse.get('Expires')));
	server.close();
});

test('cache results respect TTL', async t => {
	const {server, url} = await listen('./fixture',	{h: 'localhost', p: await port(), cache: true, cacheTTL: 1});
	const response = await get(bot, url, '/something');
	t.is(response.status, 200);
	t.false(Boolean(response.get('Expires')));
	await new Promise(resolve => {
		setTimeout(resolve, 2000);
	});
	const cachedResponse = await get(bot, url, '/something');
	t.is(cachedResponse.status, 200);
	t.false(Boolean(cachedResponse.get('Expires')));
	server.close();
});

test('gzip', async t => {
	const {server, url} = await listen('./fixture',	{h: 'localhost', p: await port(), g: true});
	const response = await getWithEncoding(human, url, '/compression.html');
	t.is(response.status, 200);
	t.true(Boolean(response.headers['content-encoding'] === 'gzip'));
	server.close();
});

test('brotli', async t => {
	const {server, url} = await listen('./fixture',	{h: 'localhost', p: await port(), b: true});
	const response = await getWithEncoding(human, url, '/compression.html');
	t.is(response.status, 200);
	t.true(Boolean(response.headers['content-encoding'] === 'br'));
	server.close();
});

test('brotli > gzip', async t => {
	const {server, url} = await listen('./fixture',	{h: 'localhost', p: await port(), g: true, b: true});
	const response = await getWithEncoding(human, url, '/compression.html');
	t.is(response.status, 200);
	t.true(Boolean(response.headers['content-encoding'] === 'br'));
	server.close();
});


test/travis-only.test.js
'use strict';

const test = require('ava');

// eslint-disable-next-line ava/no-import-test-files
const {listen, get, port} = require('./helpers');

const human = 'Chrome';

// Avoiding errors of untrusted CA: http://stackoverflow.com/questions/22654479/nodejs-https-api-testing-with-mocha-and-super-test-depth-zero-self-signed-cert
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

// test.serial('enable https devcert on develpment', async t => {
// 	const {server, url} = await listen(
// 		'./fixture',	{h: 'localhost', p: port(20), ssl: true}
// 	);
// 	const response = await get(human, url, '/something');
// 	t.is(response.text, 'some\n');
// 	server.close();
// });

test.serial('not create a dev certificate on production also with ssl specified', async t => {
	process.env.NODE_ENV = 'production';
	const {server, url} = await listen(
		'./fixture',	{h: 'localhost', p: await port(), ssl: true}
	);
	const response = await t.throwsAsync(() => get(human, url, '/something'));
	t.is(response.response, undefined);

	const responseHTTP = await get(human, url.replace('https', 'http'), '/something');
	t.is(responseHTTP.text, 'some\n');

	server.close();
});


src/cli.js
#!/usr/bin/env node

'use strict';

const meow = require('meow');
const log = require('./lib/log');
const httpServerPwa = require('.');

const IS_WIN = process.platform === 'win32';

if (IS_WIN) {
	require('readline').createInterface({
		input: process.stdin,
		output: process.stdout
	}).on('SIGINT', () => process.emit('SIGINT'));
}

process.on('SIGINT', () => {
	log.red('\nhttp-server-pwa stopped.');
	process.exit();
});

process.on('SIGTERM', () => {
	log.red('\nhttp-server stopped.');
	process.exit();
});

const cli = meow(`
	Usage
	  $ http-server-pwa [path] [options]

	Options
		-p --port       Port to use [Default: 8080]
		-h --host       Host to use [Default: localhost | Windows: 127.0.0.1]
		-f --fallback   Fallback HTML file name [Default: index.html]
		-s --https      Enable HTTPS redirect on localhost [Default: false]
		--ssl           Auto-generation SSL certificate during development [Default: false]
		-c --cache      Enable cache for Puppeteer rendering [Default: false]
		--cacheTTL      Seconds until cached content is disregarded and puppeterized again [Default: 3600 (s)].
		-d --debug      Be more verbose [Default: false]
		-g --gzip	Enable serving of gzipped files if available [Default: false]
		-b --brotli	Enable serving of brotli compressed files if available [Default: false]
		-h --help       Show this message

	Examples
	  $ http-server-pwa
	  Server started -> ./ localhost:8080
	  $ http-server-pwa dist -p 3000
	  Server started -> ./dist localhost:3000
`);

httpServerPwa(cli.input[0], cli.flags);



src/index.js
'use strict';

const {resolve, join} = require('path');
const {createServer} = require('https');
const express = require('express');
const fallback = require('express-history-api-fallback');
const {redirectToHTTPS} = require('express-http-to-https');
const certificate = require('devcert-san');

const pupperender = require('pupperender');

const loggerMiddleware = require('./lib/logger-middleware');
const log = require('./lib/log');
const {getStaticMiddleware} = require('./lib/static-middleware');

const getStHost = () => process.platform === 'win32' ? '127.0.0.1' : 'localhost';

module.exports = async (folder, options) => {
	const opt = typeof options === 'object' ? {...options} : {};
	const ROOT = resolve(folder || './');
	const PORT = opt.p || opt.port || 8080;
	const HOST = opt.h || opt.host || getStHost();
	const FALLINDEX = opt.f || opt.fallback || 'index.html';
	const DEBUG = opt.d || opt.debug || false;
	const LOCALHTTPS = opt.s || opt.https || false;
	const CACHE = opt.c || opt.cache || false;
	const CACHE_TTL = opt.cacheTTL || 3600;
	const SSL = opt.ssl || false;
	const IS_DEV = process.env.NODE_ENV !== 'production';
	const GZIP = opt.gzip || opt.g || false;
	const BROTLI = opt.brotli || opt.b || false;

	const app = express();

	// Don't redirect if the hostname is `localhost` or `127.0.0.1`
	app.use(redirectToHTTPS(LOCALHTTPS ? [] : [/localhost|127.0.0.1/]));

	app.use(loggerMiddleware(DEBUG));
	app.use(pupperender.makeMiddleware({debug: DEBUG, useCache: CACHE, cacheTTL: CACHE_TTL}));
	app.use(getStaticMiddleware(ROOT, {GZIP, BROTLI}));
	app.use(fallback(FALLINDEX, {root: ROOT}));

	const sslCert = SSL && IS_DEV ?
		await certificate.default('pwa-server', {installCertutil: true}) : null;

	return new Promise(resolve => {
		const server = sslCert ? createServer(sslCert, app) : app;
		const response = server.listen(
			PORT,
			HOST,
			() => {
				log.green('HSP started 🤘');
				log.info(
					`Path: ${ROOT} \nHost: ${sslCert ? 'https' : 'http'}://${HOST}:${PORT} \nFalling on: ${join(ROOT, FALLINDEX)}`
				);
				log.yellow('Hit CTRL-C to stop the server');
				resolve(response);
			}
		);
	});
};

